homeassistant/components/http.py

```diff
@@ -185,6 +195,41 @@ def register_path(self, method, url, callback, require_auth=True):
         """ Regitsters a path wit the server. """
         self.paths.append((method, url, callback, require_auth))
 
+    def check_expired_sessions(self):
+        """ Reemove any expired sessions. """
+        if SESSION_LOCK.acquire(False):
+            try:
+                keys = []
+                for key in self._sessions.keys():
+                    keys.append(key)
+
+                for key in keys:
+                    if self._sessions[key].is_expired:
+                        del self._sessions[key]
+                        _LOGGER.info("Cleared expired session %s", key)
+            finally:
+                SESSION_LOCK.release()
+
+    def add_session(self, key, session):
+        """ Add a new session to the list of tracked sessions """
+        self.check_expired_sessions()
+
+        if not SESSION_LOCK.acquire():
+            _LOGGER.error("Could not set session, \
+                list is locked by another thread %s", key)
+        try:
+            self._sessions[key] = session
+        finally:
+            SESSION_LOCK.release()
+
+    def get_session(self, key):
+        """ get a session by key """
+        self.check_expired_sessions()
+        session = self._sessions.get(key, None)
+        if session is not None and session.is_expired:
```

Comment:
This should never happen because we call `check_expired_sessions` two lines above.